jlm/hls/ir/hls.hpp

```diff
@@ -814,10 +816,10 @@ class load_op final : public jlm::rvsdg::simple_op
     return *util::AssertedCast<const llvm::PointerType>(&argument(0).type());
   }
 
-  [[nodiscard]] const rvsdg::valuetype &
+  [[nodiscard]] std::shared_ptr<const rvsdg::valuetype>
   GetLoadedType() const noexcept
   {
-    return *util::AssertedCast<const rvsdg::valuetype>(&result(0).type());
+    return std::dynamic_pointer_cast<const rvsdg::valuetype>(result(0).Type());
```

Comment:
I actually do not like it that you replace `AssertedCast()` with `dynamic_pointer_cast()`. The problem is that we do not have an `AssertedCast()` variant that takes a `shared_ptr`. Any chance we could introduce it and use it?

Reply:
I think we should rather get rid of the cast here entirely and store this thing as shared_ptr<const valuetype> to begin with. This requires a bit more, particularly "simple_node" is a bit too simplistic in that it squashes down everything to shared_ptr<type> from which you have to upcast again. I would like to refactor the API such that the respective classes store the "correct" types themselves and should overload a ResultType(index) instead to expose the generic form.